---
title: "Transcriptomics Exam"
author: "Sara Ometto"
date: "2022-07-3"
output:
  html_document: default
  pdf_document: default
editor_options:
  chunk_output_type: inline
  markdown:
    wrap: 72
---


# scRNAseq analysis of murine mammalian tissue

Load the required packages:

```{r message=FALSE}
library(dplyr)
library(Seurat)
library(patchwork)
library(devtools)
```

## Loading the data

The data were taken from
[PANGLAODB](https://panglaodb.se/view_data.php?sra=SRA653146&srs=SRS3044257&plot=tSNE).
It a sample of mammary tissue (mam) from mus musculus that is part of
the Tabula Muris project.

```{r}
load("SRA653146_SRS3044257.sparse.RData")
# retain only the official gene symbol
rownames(sm) <- sapply(strsplit(rownames(sm),"_"), `[`, 1)
#create the Seurat object
mam <- CreateSeuratObject(counts = sm, project = "scMammary", min.cells = 3, min.features = 200)
mam
```

## Quality control

Is a check of the principal control parameters: number of unique genes
detected in each cell (called "Features"), the total number of detected
molecules ("Count") and the percentage of reads that map to the
mitochondrial genome.

Check the presence of mithocondrial genes (mt-...) Low-quality / dying
cells often exhibit extensive mitochondrial contamination.

```{r}
grep("^mt-",rownames(mam),value = TRUE)
# add a colun with the stats
mam[["percent.mt"]] <- PercentageFeatureSet(mam, pattern = "^mt-")
```

Check the ribosomal protein genes (Rpl/Rps...)

```{r}
grep("^Rp[ls]",rownames(mam),value = TRUE)
# add a column with the stats
mam[["percent.rbp"]] <- PercentageFeatureSet(mam, pattern = "^Rp[ls]")
```

#### Visualize QC metrics

```{r}
# Feature = # genes
# Count = # detected molecules
VlnPlot(mam, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rbp"), ncol = 4, cols = "mediumpurple1")

# plot without dots:
VlnPlot(mam, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rbp"), ncol = 4, pt.size=0, cols = "mediumpurple1")
```

Then I check if there is a correlation between the different parameters
with FeatureScatter plots:

-   Correlation between % of mitochondrial RNA and number of reads

-   Correlation between number of genes and number of reads

-   Correlation between % of rRNA and number of reads


```{r}
FeatureScatter(mam, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(mam, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
FeatureScatter(mam, feature1 = "nCount_RNA", feature2 = "percent.rbp")
```

Low-quality cells or empty droplets will often have very few genes,
while cell doublets or multiplets exhibit an aberrantly high gene count.
The total number of molecules detected within a cell correlates strongly
with unique genes, is the only visible correlation.


#### QC threshold

From these plots I decide a threshold for the number of genes:
between 200 and 4000. The mithocondrial percentage should be lower than 5%.

```{r}
mam <- subset(mam, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 5)
mam
```
The remaining samples are 3134 (from 3454 of before)

#### Data Normalization
10x data are usually just transformed into counts per million (or counts x 10,000 reads) 
The final “expression estimate” it’s given by the log of the normalized counts:
```{r}
mam <- NormalizeData(mam, normalization.method = "LogNormalize", scale.factor = 10000)
```

## Cell cycle effect

In order to assign a CC phase to our cells, that come from mouse, 
I needed to create a list of putative genes suitable for this animal. 

```{r message=FALSE, warning=FALSE}
convertHumanGeneList <- function(x){
  require("biomaRt")
  human = useMart(biomart="ensembl", dataset = "hsapiens_gene_ensembl", 
                  verbose = TRUE, host = "https://dec2021.archive.ensembl.org")
  mouse = useMart(biomart="ensembl", dataset = "mmusculus_gene_ensembl", 
                  verbose = TRUE, host = "https://dec2021.archive.ensembl.org")
  genes = getLDS(attributes = c("hgnc_symbol"), 
                 filters = "hgnc_symbol", values = x ,
                 mart = human, attributesL = c("mgi_symbol"), 
                 martL = mouse, uniqueRows=T)
  
  humanx <- unique(genes[, 2])
  return(humanx)
}
m.s.genes <- convertHumanGeneList(cc.genes.updated.2019$s.genes)
m.g2m.genes <- convertHumanGeneList(cc.genes.updated.2019$g2m.genes)

mam <- CellCycleScoring(mam, s.features = m.s.genes, g2m.features = m.g2m.genes, set.ident = TRUE)
```

## Subset and scaling

I create a **subset**, keeping only the 2000 most variable genes.

```{r}
mam <- FindVariableFeatures(mam, selection.method = "vst", nfeatures = 2000)
?FindVariableFeatures
# top 10 most variable genes:
top10 <- head(VariableFeatures(mam), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(mam)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```


I shift the expression of each gene so that the mean is 0 and the variance is 1 

```{r}
all.genes <- rownames(mam)
mam <- ScaleData(mam, features = all.genes)
```

## Dimensional reduction

I perform PCA on the 2000 most variable genes

```{r}
mam <- RunPCA(mam, features = VariableFeatures(object = mam))
print(mam[["pca"]], dims = 1:5, nfeatures = 5)

# visualize the most variable genes
VizDimLoadings(mam, dims = 1:2, reduction = "pca")
VizDimLoadings(mam, dims = 3:4, reduction = "pca")
VizDimLoadings(mam, dims = 4:5, reduction = "pca")
```

#### Visualization

Then I visualize the projection of the cells in the first two PCs:
(Cells are colored according to cell cycle phase)

```{r}
DimPlot(mam, reduction = "pca")
```

The cell cycle seems to be ininfluent!

#### Select number of PCs

```{r}
ElbowPlot(mam, ndims= 40)
```


------------------------
Choosing how many dimensions to use can vary depending on the method we
choose. In general it's better to keep all PC until 70/75% of the
variance is explained

Rule of the thumb:
```{r}
pc.touse <- (mam$pca@stdev)^2
pc.touse <- pc.touse/sum(pc.touse)
pc.touse <- cumsum(pc.touse)[1:50]
pc.touse <- min(which(pc.touse>=0.85))
pc.touse
```

From this the principal components that retain 75-80% of the variance are 22.
I chose to do the analysis for 22 PCs and 15 PCs.

## Clustering with 15 PC

The first step uses the FindNeighbors function, which constructs a KNN graph based on the euclidean distance in PCA space and refines the edge weights using the Jaccard similarity

```{r}
mam15 <- FindNeighbors(mam, dims = 1:15)
```

To cluster the cell we use the FindClusters function, which uses the
Louvain algorithm to iteratively group cells together

```{r}
mam15 <- FindClusters(mam15, resolution = 0.3)
# 9 clusters
```


#### TSNE

I plot the clusters using T Stochastic Neighbor Embedding

```{r}
mam15.tsne <- RunTSNE(mam15, dims=1:15)
DimPlot(mam15.tsne, reduction = "tsne")
```

#### UMAP

Uniform Manifold Approximation and Projection is generally preferred 
but requires the installation of a new package

```{r message=FALSE, warning=FALSE}
mam15.UMAP <- RunUMAP(mam15, dims = 1:15)
DimPlot(mam15.UMAP, reduction = "umap")
```

We can also check whether some of the critial quality paramteres influenced the clustering we got

```{r}
VlnPlot(mam15.UMAP,features="nCount_RNA")
VlnPlot(mam15.UMAP,features="nFeature_RNA")
VlnPlot(mam15.UMAP,features="percent.mt")
VlnPlot(mam15.UMAP,features="percent.rbp")
```

or the cell cycle

```{r}
library(ggplot2)
library(dbplyr)

mam15@meta.data %>%
  group_by(seurat_clusters,Phase) %>%
  count()  %>% 
  group_by(seurat_clusters) %>%
  mutate(percent=100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=Phase)) +
  geom_col() + 
  ggtitle("Percentage of cell cycle phases per cluster")
```

For all the other quality check the clustering seems reasonable.

## Clustering with 22 PC

```{r}
mam22 <- FindNeighbors(mam, dims = 1:22)
```

To cluster the cell we use the FindClusters function.

```{r}
mam22 <- FindClusters(mam22, resolution = 0.5)
# 12 clusters
```

#### TSNE

I plot the clusters using T Stochastic Neighbor Embedding

```{r}
mam22.tsne <- RunTSNE(mam22, dims=1:22)
DimPlot(mam22.tsne, reduction = "tsne")
```

#### UMAP

Uniform Manifold Approximation and Projection is generally preferred 
but requires the installation of a new package

```{r message=FALSE}
mam22.UMAP <- RunUMAP(mam22, dims = 1:22)
DimPlot(mam22.UMAP, reduction = "umap")
```



We can also check whether some of the critical quality parameters influenced the clustering we got

```{r}
VlnPlot(mam22.UMAP,features="nCount_RNA")
VlnPlot(mam22.UMAP,features="nFeature_RNA")
VlnPlot(mam22.UMAP,features="percent.mt")
VlnPlot(mam22.UMAP,features="percent.rbp")
```

 or the cell cycle

```{r}

mam22@meta.data %>%
  group_by(seurat_clusters,Phase) %>%
  count()  %>% 
  group_by(seurat_clusters) %>%
  mutate(percent=100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=Phase)) +
  geom_col() + 
  ggtitle("Percentage of cell cycle phases per cluster")
```
Maybe the quantity of mt RNA in cluster 4 could make 
the identification of marker genes difficult.
The quality check is good also with 22 PCs.



## Finding marker genes for 15PCs clustering

Seurat also includes a function that can be used to find genes over
expressed between two clusters or overexpressed in one cluster with
respect to all the others

The one vs all clusters analyses can be iterated automatically, 
and we can output the top n (in this case 5) genes for each cluster. 

Notice that here they are sorted by logFC - more informative than “p_val_adj”, 
since a lot of genes will have a FDR close to zero with smallest changes:

```{r message=FALSE}
mam15 <- mam15.UMAP
mam15.markers <- FindAllMarkers(mam15, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

mam15.markers %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)

```

### Heatmap

```{r}
mam15.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(mam15, features = top10$gene) + NoLegend()
```

Looking at the Heatmap we can see that clusters 0 and 4 show quite some
of similarities. Also clusters 5, 7 and 8 are similar.


### Going in depth
In real life, having two clusters with “too similar” patterns of DE/marker genes might be a problem. 
Can we conclude that after all they are the same cell type, partitioned wrongly into two separate clusters, or they are indeed two different cell types or subtypes, with a few genes “making the difference” from one another? 


#### T cells - 0/4

We want to asses the cell type shared by cluster 0 and 4.

```{r}
c0and4.markers <- FindMarkers(mam15, ident.1 = c(0,4), min.pct = 0.25, test.use = "wilcox")
c0and4.markers <- c0and4.markers[order(-c0and4.markers$avg_log2FC), ] 
head(c0and4.markers, n = 20)
```

Cd3d, Cd3g, Ms4a4b, **Trbc2** and 1 (t cell receptor beta), Trac (T cell receptor alfa) - Tcell
It seems that they are all **T cells**

Let's see the genes making the difference between those two clusters:

Genes overexpressed in cluster 0 vs cluster 4

```{r}
c0vs4.markers <- FindMarkers(mam15, ident.1 = 0, ident.2 = 4, min.pct = 0.25, test.use = "wilcox") 
c0vs4.markers <- c0vs4.markers[order(-c0vs4.markers$avg_log2FC),] 
head(c0vs4.markers, n = 30)
```
Dapl1 - death associated protein 1 - T cells activates CD8

Lef1 - lymphoid enhancer factor (regulates T cell receptor) - CD8 T cells 

Cd8b1 - CD8 antigen, beta chain 1 - CD8 T cells 

Rapgef6 - Rap guanine nucleotide exchange factor (GEF) 6 transmission signal - CD8 T cells 

Cluster 0: **Cytotoxic T CELLS**


And overexpressed in cluster 4 vs cluster 0

```{r}
c4vs0.markers <- FindMarkers(mam15, ident.1 = 4, ident.2 = 0, min.pct = 0.25, test.use = "wilcox") 
c4vs0.markers <- c4vs0.markers[order(-c4vs0.markers$avg_log2FC),] 
head(c4vs0.markers, n = 40)
```
Ccl5 - chemokine ligand - all the cells (overexpressed in activated T)

Tnfrsf4 and 9 - TNF receptor superfamily (citokyne receptor) - 4 overexpressed in T cells

**Il2rb** - expressed mainly in T reg cells

Icos -  inducible T cell co-stimulator  or Il2rb interleukin 2 receptor for T reg

Ctla4 - citotoxic T cell

Cluster 4: **Regulatory T CELLS**

```{r}
c4.markers <- FindMarkers(mam15, ident.1 = 4, min.pct = 0.25, test.use = "wilcox")
c4.markers<- c4.markers[order(-c4.markers$avg_log2FC),] 
head(c4.markers, n = 20)
```


#### Macrophages - 5

Cluster 5, 7 and 8
```{r}
c5and6and7.markers <- FindMarkers(mam15, ident.1 = c(5,6,7), min.pct = 0.25, test.use = "wilcox")
c5and6and7.markers <- c5and6and7.markers[order(-c5and6and7.markers$avg_log2FC), ] 
head(c5and6and7.markers, n = 10)
```
Krt14 and Krt17 basal cells,
Lyz2 macrophages,
Plet1 al biological processes? - basal cells luminal epitelial,
Slpi basal cells,
Sfn basal and luminal epitelial,
Acta2 and Cxcl14 basal

seem to be basal cell, but I need to further assess the gene expression

Cluster 5
```{r}
c5vs78.markers <- FindMarkers(mam15, ident.1 = 5, ident.2 = c(7,8), min.pct = 0.25, test.use = "wilcox") 
c5vs78.markers <- c5vs78.markers[order(-c5vs78.markers$avg_log2FC),] 
head(c5vs78.markers , n = 10)
```
Lyz2, Ccl4, C1qb - ONLY expressed in macrophages

H2-Ab1 - b cells and macrophages



Cluster 5 vs ALL:
**Lyz2** -> produces Lyzozyme which has bacteriolytic function, contained in macrophages

**Ccl4**, **IL-1β** expressed in a wide range of cells, especially in macrophages

**MACROPHAGES**

#### Luminal epithelial - 7/8
Cluster 7
```{r}
c7vs8.markers <- FindMarkers(mam15, ident.1 = 7, ident.2 = 8, min.pct = 0.25, test.use = "wilcox") 
c7vs8.markers <- c7vs8.markers[order(-c7vs8.markers$avg_log2FC),] 
head(c7vs8.markers , n = 10)
```
Csn3 - luminal epithelial (caseina) - milk gene

Lcn2 lipocalin 2 - luminal progenitor

Plet1 - luminal cells

Wfdc18 - WAP four-disulfide core domain 18 - milk gene in the luminal progenitor

Aldh1a3 - progenitor luminal marker (from article)


Cluster 7 vs ALL:
Csn3 produces caseine in luminal epithelial cells, **Wfdc18** WAP is produced 
selectively by committed luminal cells within mammary ducts and alveoli 

**LUMINAL EPITHELIAL PROGENITOR**


Cluster 8
```{r}
c8vs7.markers <- FindMarkers(mam15, ident.1 = 8, ident.2 = 7, min.pct = 0.25, test.use = "wilcox") 
c8vs7.markers<- c8vs7.markers[order(-c8vs7.markers$avg_log2FC),] 
head(c8vs7.markers , n = 20)
```
Ptn - pleiotropin, expressed in epithelial cells of the mammary gland

Amphiregulin (AREG), a ligand for epidermal growth factor receptor, 
is required for mammary gland ductal morphogenesis - found in mature luminal

Ly6a marker for mature luminal cells

Prlr - prolactine receptor


```{r}
cluster8.markers <- FindMarkers(mam15, ident.1 = 8, min.pct = 0.25, test.use = "wilcox")
cluster8.markers<- cluster8.markers[order(-cluster8.markers$avg_log2FC),] 
head(cluster8.markers, n = 20)
```

**Areg**,  Prlr prolactin receptor

Pgr progesterone receptor - Mature luminal (from tabula muris article)

**MATURE LUMINAL**


#### B cells - 1

Cluster 1 vs all
```{r}
cluster1.markers <- FindMarkers(mam15, ident.1 = 1, min.pct = 0.25, test.use = "wilcox")
cluster1.markers<- cluster1.markers[order(-cluster1.markers$avg_log2FC),] 
head(cluster1.markers, n = 20)
```

**Igkc** immunoglobulin kappa constant - **B cells**

Cd74 - B cells and macrophages

Cd79a - B cells creates Ig alpha

Cd79b - B cells creates Ig beta

H2-.. - histocompatibility complex 2, found on B cells


#### Stromal cells - 2
```{r}
cluster2.markers <- FindMarkers(mam15, ident.1 = 2, min.pct = 0.25, test.use = "wilcox")
cluster2.markers<- cluster2.markers[order(-cluster2.markers$avg_log2FC),] 
head(cluster2.markers, n = 20)
```

**Dcn** decorin (proteoglycan) - **Stromal cells**

Gsn gelsolina (actin binding protein) - Stromal cells

Tnfaip6 - stromal cells

Col3a1 collagene type III - Stromal cells

Col1 collagen type II - Stromal cells



#### Endotelial cells - 3

```{r}
cluster3.markers <- FindMarkers(mam15, ident.1 = 3, min.pct = 0.25, test.use = "wilcox")
cluster3.markers<- cluster3.markers[order(-cluster3.markers$avg_log2FC),] 
head(cluster3.markers, n = 20)
```

Fabp4 fatty acid biding protein - endotelial cells

**Cldn5** endothelial thight junctions - **endotelial cells**

Aqp1 aquaporin 1 - endotelial cells

Cd36 lipid transport - endotelial cells

Esam - endotelial cell specific adhesion protein

#### Basal cells - 6
Cluster 6
```{r}
c6vs8.markers <- FindMarkers(mam15, ident.1 = 6, ident.2 = 8, min.pct = 0.25, test.use = "wilcox") 
c6vs8.markers <- c6vs8.markers[order(-c6vs8.markers$avg_log2FC),] 
head(c6vs8.markers , n = 10)
```
**Krt14** and Krt17 marker for basal cells, also Acta2

**BASAL CELLS**




## Visualizing marker genes

Plot the expression of the markers with a heatmap

```{r}
FeaturePlot(mam15, features = c("Trbc2", "Cd79a", "Dcn", "Fabp4", "Lyz2", "Krt14", "Plet1", "Areg"))
```

I inspect in depth the subpopulations of T cells:

```{r}
FeaturePlot(mam15, features = "Il2rb", repel = T) 
FeaturePlot(mam15, features = "Cd8b1", repel = T) 
```

Or with a violin plot

```{r}
VlnPlot(mam15, features = "Cd8b1")
VlnPlot(mam15, features = "Cd79a")
VlnPlot(mam15, features = "Dcn")
VlnPlot(mam15, features = "Fabp4")
VlnPlot(mam15, features = "Il2rb")
VlnPlot(mam15, features = "Lyz2")
VlnPlot(mam15, features = "Krt14")
VlnPlot(mam15, features = "Plet1")
VlnPlot(mam15, features = "Areg")
```

We can visualize all the marker genes and their expression (CPM) in each
cluster using a dot plot.

```{r}
library(ggrepel)
DotPlot(mam15, features = c("Trbc2", "Cd79a", "Dcn", "Fabp4", "Il2rb", "Lyz2", "Krt14", "Plet1", "Areg")) + theme(axis.text.x = element_text(angle = 90))
```
Dot plot with Cd8b1 gene marker for cluster 0 instead of Trbc2:

```{r}
library(ggrepel)
DotPlot(mam15, features = c("Cd8b1", "Cd79a", "Dcn", "Fabp4", "Il2rb", "Lyz2", "Krt14", "Plet1", "Areg")) + theme(axis.text.x = element_text(angle = 90))
```
Both cluster 0 and 4 are T cells, but it seems that only cluster 0 contains CD8+ T cells.

# **Final Result**

Using marker genes to infere the subtypes of each cluster was quite
difficult and needed a deep knowlegde of the fild (which in our case is
limited). Eventually we were able to identify 10 different cells types
of which 4 subtypes of ODC, 2 subtypes of AST and Microglia, Neuros, OPC
ad Endothelial cells.

Comparing the results with the automated pipeline employed by Panglao,
we were able to label and cluster the cells with unknown cell type. Also
we did not found Pancreatich stellate cells and Pericytes, may be due to
different clustering parameters.

```{r}
new.cluster.ids <- c("T cytotoxic", "B cells", "Stromal cells", "Endotelial cells", "T reg", "Macrophages", "Basal cells", "Luminal Progenitor ", 'Mature Luminal') 
names(new.cluster.ids) <- levels(mam15) 
MAM15<- RenameIdents(mam15, new.cluster.ids) 
DimPlot(MAM15, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()

```

DimPlot(mam15.tsne, reduction = "tsne")
```{r}
names(new.cluster.ids) <- levels(mam15.tsne) 
MAM15.TSNE<- RenameIdents(mam15.tsne, new.cluster.ids) 
DimPlot(MAM15.TSNE, reduction = "tsne", label = TRUE, pt.size = 0.5) + NoLegend()
```

